src/colony/modules/_net_proxied.js

```diff
@@ -0,0 +1,200 @@
+    if (this._tunnel) return cb(null, this._tunnel);
+    
+    var self = this;
+    if (!this._pending) createTunnel(function (e, tunnel) {
+      delete self._pending;
+      if (e) return self.emit('tunnel', e);
+      
+      self._tunnel = tunnel;
+      tunnel.on('close', function () {
+        self._tunnel = null;
+      });
+      var streamProto = Object.create(ProxiedSocket.prototype);
+      streamProto._tunnel = tunnel;
+      tunnel._streamProto = streamProto;
+      self.emit('tunnel', null, tunnel);
+    });
+    this._pending = true;
+    this.once('tunnel', cb);
+};
+
+var local_matchers = PROXY_LOCAL.split(' ').map(function (str) {
+  var parts = str.split('/');
+  if (parts.length > 1) {
+    // IPv4 + mask
+    var bits = +parts[1],
+        mask = 0xFFFFFFFF << (32-bits) >>> 0,
+        base = net._ipStrToInt(parts[0]) & mask;      // NOTE: left signed to match test below
+    return function (addr, host) {
+      return ((addr & mask) === base);
+    };
+  } else if (str[0] === '.') {
+    // base including subdomains
+    str = str.slice(1);
+    return function (addr, host) {
+      var idx = host.lastIndexOf(str);
+      return (~idx && idx + str.length === host.length);
+    };
+  } else return function (addr, host) {
+    // exact domain/address 
+    return (host === str);
+  }
+});
+
+function protoForConnection(host, port, opts, cb) {   // CAUTION: syncronous callback!
+  var addr = (net.isIPv4(host)) ? net._ipStrToInt(host) : null,
+      force_local = !PROXY_TOKEN || (opts._secure && !PROXY_TRUSTED) || (opts.proxy === false),
+      local = force_local || local_matchers.some(function (matcher) { return matcher(addr, host); });
+  if (_PROXY_DBG) {
+      if (force_local) console.log(
+        "Forced to use local socket to \"%s\". [token: %s, secure/trusted: %s/%s, opts override: %s]",
+        host, Boolean(PROXY_TOKEN), Boolean(opts._secure), Boolean(PROXY_TRUSTED), (opts.proxy === false)
+      );
+      else console.log("Proxied socket to \"%s\"? %s", host, !local);
+  }
+  if (local) cb(null, net._CC3KSocket.prototype);
+  else tunnelKeeper.getTunnel(function (e, tunnel) {
+    if (e) return cb(e);
+    cb(null, tunnel._streamProto);
+  });
+}
+
+/**
+ * ProxiedSocket
+ */
+
+function ProxiedSocket(opts) {
+  if (!(this instanceof ProxiedSocket)) return new ProxiedSocket(opts);
+  net.Socket.call(this, opts);
+  this._tunnel = this._opts.tunnel;
+  this._setup(this._opts);
+}
+util.inherits(ProxiedSocket, net.Socket);
+
+ProxiedSocket.prototype._setup = function () {
+  var type = (this._secure) ? 'tls' : 'net';
+  this._transport = this._tunnel.createStream(type);
+  
+  var self = this;
+  // TODO: it'd be great if we is-a substream instead of has-a…
+  this._transport.on('data', function (d) {
+    var more = self.push(d);
+    if (!more) self._transport.pause();
+  });
+  this._transport.on('end', function () {
+    self.push(null);
```

Comment:
Not quite sure this is what you're asking, but ProxiedSocket may never emit a `'close'` event…investigating.